### PR TITLE
Minor fixes and refactoring

### PR DESCRIPTION
**Pull request by madoka-wizard**
- Minor changes and code style fixes
- Implemented amazing testing system (see `WhenAny` tests)

Detailed description:
- Add branch coverage to lcov report
- Update .clang-tidy rules to avoid stupid warnings
- Fix stupid GCC warning
- Make style_guide more comprehensive
- Add MBkkt's TODO
- Implement ranged `Wait`/`WaitFor`/`WaitUntil` for iterators
- Fix Wait tests and update test/CMakeLists.txt
- Update `WhenAny` implementation
- Make ultimate `WhenAny` tests using cartesian product of different params
- Add `InlineCore`
- Separate `WhenAll` implementation from interface
- Code style fixes
- `WaitGroup` fixes

### CODE REVIEW

File: .github/workflows/test_coverage.yml [hunk 1]
```diff
@@ -52,11 +52,11 @@ jobs:
       - name: Collect lcov info
         run:
           cd build &&
-          lcov -d . -c                  -o lcov.info &&
-          lcov -r lcov.info '/usr/*'    -o lcov.info &&
-          lcov -r lcov.info '*/_deps/*' -o lcov.info &&
-          lcov -r lcov.info '*/test/*'  -o lcov.info &&
-          lcov --list lcov.info
+          lcov --rc lcov_branch_coverage=1 -d . -c                  -o lcov.info &&
```
Comment (MBkkt) [inline .github/workflows/test_coverage.yml:55]: Make two lcov.info one for codecov (without branch), one for us (with branch)

File: CMakeLists.txt [hunk 2]
```diff
@@ -26,7 +26,7 @@ set(CMAKE_C_EXTENSIONS OFF)
 set(CMAKE_C_VISIBILITY_PRESET hidden)
 
 set(CMAKE_VISIBILITY_INLINES_HIDDEN ON)
-set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+set(CMAKE_POSITION_INDEPENDENT_CODE ON) # TODO(MBkkt) Do we really need this?
```
Comment (MBkkt) [inline CMakeLists.txt:29]: I think we can remove this and LTO

File: include/yaclib/algo/detail/wait_impl.hpp [hunk 3]
```diff
@@ -39,6 +41,37 @@ bool Wait(const Timeout& t, Cores&... cs) {
   return ready;
 }
 
+template <typename Timeout, typename... Cores>
+bool Wait(const Timeout& t, Cores&... cs) {
+  static_assert(sizeof...(cs) >= 1, "Number of futures must be at least one");
+  static_assert((... && std::is_same_v<detail::BaseCore, Cores>), "Futures must be Future in Wait function");
+  auto range = [&](auto&& functor) {
+    return (... & functor(cs));  // TODO(MBkkt) Maybe (functor(cs) & ...) is better
```
Comment (MBkkt) [inline include/yaclib/algo/detail/wait_impl.hpp:49]: - [ ] Remove TODO

These are the same expressions, the only difference is in associativity, which doesn't matter here and is correct
(... & some) -- (somef1 & somef2) & somef3
(some & ...) -- somef1 & (somef2 & somef3)

File: include/yaclib/algo/detail/wait_impl.hpp [hunk 4]
```diff
@@ -39,6 +41,37 @@ bool Wait(const Timeout& t, Cores&... cs) {
...
+}
+
 extern template bool Wait<NoTimeoutTag, BaseCore>(const NoTimeoutTag&, BaseCore&);
 
+template <typename Timeout, typename ValueIt, typename RangeIt>
+bool Wait(const Timeout& t, ValueIt it, RangeIt begin, RangeIt end) {
+  static_assert(util::IsFutureV<typename std::iterator_traits<ValueIt>::value_type>,
+                "Wait function Iterator must be point to some Future");
+  if (begin == end) {
+    return true;
```
Comment (MBkkt) [inline include/yaclib/algo/detail/wait_impl.hpp:61]: Please add test

File: include/yaclib/algo/detail/when_any_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+template <typename T, WhenPolicy P>
+using AnyCombinatorPtr = util::Ptr<AnyCombinator<T, P>>;
+
+template <WhenPolicy P, typename T, typename ValueIt, typename RangeIt>
+Future<T> WhenAnyImpl(ValueIt it, RangeIt begin, RangeIt end) {
+  if (begin == end) {
+    return {};
+  }
+  if (auto next = begin; ++next == end) {
+    return std::move(*it);
```
Comment (MBkkt) [inline include/yaclib/algo/detail/when_any_impl.hpp:138]: Add test with vector length one

File: include/yaclib/algo/detail/when_all_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    if constexpr (!std::is_void_v<T> && !IsArray) {
+      AllCombinatorBase<FutureValue>::_results.resize(size);
+    }
+  }
+
+  Promise<FutureValue> _promise;
+};
+
+template <size_t N, typename T, typename... Ts>
+void WhenAllImpl(util::Ptr<AllCombinator<T, N>>& combinator, Future<T>&& head, Future<Ts>&&... tail) {
```
Comment (MBkkt) [inline include/yaclib/algo/detail/when_all_impl.hpp:102]: Add constexpr plz

File: include/yaclib/algo/detail/when_any_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  for (; begin != end; ++begin) {
+    it->GetCore()->SetCallbackInline(combinator);
+    std::move(*it).Detach();
+    ++it;
+  }
+  return std::move(future);
+}
+
+template <WhenPolicy P, typename T, typename... Ts>
+void WhenAnyImpl(AnyCombinatorPtr<T, P>& combinator, Future<T>&& head, Future<Ts>&&... tail) {
```
Comment (MBkkt) [inline include/yaclib/algo/detail/when_any_impl.hpp:160]: Add constexpr plz

File: include/yaclib/algo/when_all.hpp [hunk 8]
```diff
@@ -1,123 +1,19 @@
...
  * \tparam T type of all passed futures
  * \param begin, size the range of futures to combine
- * \return Future<vector<T>>
+ * \return Future<std::vector<T>>
  */
-template <typename It, typename T = util::detail::FutureValueT<typename std::iterator_traits<It>::value_type>>
+template <WhenPolicy P = WhenPolicy::FirstFail, typename It,
+          typename T = util::detail::FutureValueT<typename std::iterator_traits<It>::value_type>>
 auto WhenAll(It begin, size_t size) {
   auto [future, combinator] = detail::AllCombinator<T>::Make(size);
```
Comment (MBkkt) [inline include/yaclib/algo/when_all.hpp:18]: Add for all WhenAll static_assert(p == WhenPolicy::FirstFail, "TODO(Ri7ay) Add other policy for WhenAll")

File: include/yaclib/async/detail/wait_core.hpp [hunk 9]
```diff
@@ -24,7 +24,7 @@ class WaitCore : public util::IRef {
 
   template <typename Clock, typename Duration>
   bool Wait(std::unique_lock<std::mutex>& guard, const std::chrono::time_point<Clock, Duration>& timeout_time) {
-    return cv.wait_until(guard, timeout_time, [&] {
+    return cv.wait_until(guard, timeout_time, [this] {
```
Comment (MBkkt) [inline include/yaclib/async/detail/wait_core.hpp:27]: remove this diff

File: include/yaclib/util/result.hpp [hunk 10]
```diff
@@ -74,14 +76,13 @@ class Result {
   static_assert(!std::is_same_v<T, std::error_code>, "Result cannot be instantiated with std::error_code");
   static_assert(!std::is_same_v<T, std::exception_ptr>, "Result cannot be instantiated with std::exception_ptr");
 
-  struct Unit {};
-  using VariantT =
-      std::variant<std::monostate, std::error_code, std::exception_ptr, std::conditional_t<std::is_void_v<T>, Unit, T>>;
+  using VariantT = std::variant<std::monostate, std::error_code, std::exception_ptr,
```
Comment (MBkkt) [inline include/yaclib/util/result.hpp:79]: using ValueT = std::conditional_t<std::is_void_v<T>, detail::Unit, T>;

File: include/yaclib/util/result.hpp [hunk 11]
```diff
@@ -113,35 +114,32 @@ class Result {
...
-      }
-      case ResultState::Empty: {
+      default:
         throw ResultEmpty{};
-      }
     }
   }
 
   [[nodiscard]] ResultState State() const noexcept {
     if (std::holds_alternative<std::monostate>(_result)) {
```
Comment (MBkkt) [inline include/yaclib/util/result.hpp:133]: I think better if first
if (std::holds_alternative<ValueT>(_result)) {
}
then exception, then error, last empty

File: src/algo/wait_group.cpp [hunk 12]
```diff
@@ -6,6 +6,13 @@ WaitGroup::WaitGroup() {
   _callback.IncRef();
 }
 
+template void WaitGroup::Add<void>(Future<void>&);
+
+void WaitGroup::Add(Future<void>&& f) {
```
Comment (MBkkt) [inline src/algo/wait_group.cpp:11]: Add test

**Review by MBkkt — CHANGES_REQUESTED** [review]

**Comment by kononovk** [comment]
Merged this changes in #123

**Pull request closed without merging**